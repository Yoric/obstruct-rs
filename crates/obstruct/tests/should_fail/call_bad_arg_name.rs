#![feature(associated_const_equality)]
fn struct_bad_field_names() {
    // Check that the code won't build if the field names are incorrect

    const RED: &str = "red";
    const GREEN: f64 = 1.0;
    const BLUE: () = ();

    use obstruct_macros::{instruct, destruct};

    let structured = instruct! { red: RED, green: GREEN, blue: BLUE };

    destruct! { let {red, green, oops} = structured }; // Look, we've used oops!
    // This fails with an error message mentioning oops.
}
fn main() {}