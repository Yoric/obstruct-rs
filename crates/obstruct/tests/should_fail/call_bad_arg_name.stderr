error[E0271]: type mismatch resolving `<blue<()> as Field<()>>::NAME == "green"`
  --> tests/should_fail/call_bad_arg_name.rs:13:5
   |
13 |     destruct! { let {red, green, oops} = structured }; // Look, we've used oops!
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `"green"`, found `"blue"`
   |
   = note: expected constant `"green"`
              found constant `"blue"`
note: required by a bound in `struct_bad_field_names::assert_type`
  --> tests/should_fail/call_bad_arg_name.rs:13:5
   |
13 |     destruct! { let {red, green, oops} = structured }; // Look, we've used oops!
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_type`
   = note: this error originates in the macro `destruct` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<green<f64> as Field<f64>>::NAME == "oops"`
  --> tests/should_fail/call_bad_arg_name.rs:13:5
   |
13 |     destruct! { let {red, green, oops} = structured }; // Look, we've used oops!
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `"oops"`, found `"green"`
   |
   = note: expected constant `"oops"`
              found constant `"green"`
note: required by a bound in `struct_bad_field_names::assert_type`
  --> tests/should_fail/call_bad_arg_name.rs:13:5
   |
13 |     destruct! { let {red, green, oops} = structured }; // Look, we've used oops!
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_type`
   = note: this error originates in the macro `destruct` (in Nightly builds, run with -Z macro-backtrace for more info)
